### ISSUE TITLE
Better Solver

### ISSUE THREAD

**Issue by Possseidon** [Possseidon/mines3#6]
## Current Solver

The current solver always looks at a full chain of connected unrevealed fields and tries all possible combinations of mines. The complexity is the total number of unrevealed fields that have to be looked at in parallel. While this worked surprisingly well, it can lead to high complexity fairly quickly.

## New Solver

- For each revealed field:
  - Create a list of all sets of mines that are valid
    - I.e. if there is one more mine left, a list where each set contains exactly one of the remaining fields
    - Only one valid list -> reveal/flag all
- As an optimization, already fully separate fields that do not have any overlap before going to the next step
- For k := 2 to min(number of revealed fields with at least one unrevealed field next to it, max_complexity)
  - Find all combinations of k fields
    - Filter out combinations that do not have any overlapping fields
    - For k > 2 this overlap can go through multiple fields
      - i.e. `[A, B] [B, C] [C, D]` counts as overlap even though `[A, B]` and `[C, D]` do not overlap
      - but `[A, B] [B, C] [D, E]` does not, since `[D, E]` does not share any common fields with the rest
      - likewise `[A, B] [B, C] [D, E] [E, F]` does not either, since it results in two separate sets `[A, B, C]` and `[D, E, F]`
    - For all fields of the combination, construct the cartesian product between all their lists
      - Filter out those with mismatched mine-placements
        - E.g. if `B` is a mine in `[A, B]` but isn't a mine in `[B, C]`, discard it.
      - If a field contains a mine in all possible configurations, flag it
      - If a field contains no mine in all possible configurations, reveal it
      - Ignore other fields

**Comment by Possseidon** [comment]
Similar approach with some major improvements:

- While let Some(entry) = pop
    - Find one that overlaps
    - swap remove it
    - cartesian product between their k-combinations
    - pre-calculate the number of elements in this set using some basic math
        - if it's too big; mark the required complexity and continue
    - Filter out the ones where the overlap is not the same
    - There has to be at least one match
    - If it's just one, apply it; don't break though, continue on with the next step!
    - find another one that overlaps and check all of its k-combinations that match
    - multiple the number of k-combinations on top of the previous product
        - if it's too big; ...
    - If there's only one, apply it; don't break though, continue on
    - If there's multiple ones, find another one and check all of its k-combinations
        - etc...
    - If there's no more overlap but still more than one item, break
        - nothing we can do with this set of fields yet

**Comment by Possseidon** [comment]
More random thoughts:

```
1 1 1 0 0 = 2
0 1 1 1 0 = 2
0 0 1 1 1 = 2
```

- Take the first row.
- If the row is "= 0" (should never happen) all 1 can be marked as 0.
- If the row is "= popcount" all 1 can be marked as 1.
- If there is no other row with overlap, any column that is all 0 or 1 is solved and the row can be discarded.
- If there is a row that is a subset of this row:
  - Remove all 1s in the subset from the superset and subtract count.
- Find the row with the largest overlap (bitand + popcount).
- Split it into three parts, only left, only right, overlap
- Find list of possible k numbers
- Generate all possible k combinations
